ddtrace/contrib/aioredis/patch.py

```diff
@@ -0,0 +1,175 @@
+
+
+def patch():
+    if getattr(aioredis, "_datadog_patch", False):
+        return
+    setattr(aioredis, "_datadog_patch", True)
+    pin = Pin()
+    if aioredis_version >= (2, 0):
+        _w("aioredis.client", "Redis.execute_command", traced_execute_command)
+        _w("aioredis.client", "Redis.pipeline", traced_pipeline)
+        _w("aioredis.client", "Pipeline.execute", traced_execute_pipeline)
+        pin.onto(aioredis.client.Redis)
+    else:
+        _w("aioredis", "Redis.execute", traced_13_execute_command)
+        _w("aioredis", "Redis.pipeline", traced_13_pipeline)
+        _w("aioredis.commands.transaction", "Pipeline.execute", traced_13_execute_pipeline)
+        pin.onto(aioredis.Redis)
+
+
+def unpatch():
+    if not getattr(aioredis, "_datadog_patch", False):
+        return
+
+    setattr(aioredis, "_datadog_patch", False)
+    if aioredis_version >= (2, 0):
+        _u(aioredis.client.Redis, "execute_command")
+        _u(aioredis.client.Redis, "pipeline")
+        _u(aioredis.client.Pipeline, "execute")
+    else:
+        _u(aioredis.Redis, "execute")
+        _u(aioredis.Redis, "pipeline")
+        _u(aioredis.commands.transaction.Pipeline, "execute")
+
+
+async def traced_execute_command(func, instance, args, kwargs):
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return await func(*args, **kwargs)
+
+    decoded_args = [arg.decode() if isinstance(arg, bytes) else arg for arg in args]
+    with _trace_redis_cmd(pin, config.aioredis, instance, decoded_args):
+        return await func(*args, **kwargs)
+
+
+async def traced_pipeline(func, instance, args, kwargs):
+    pipeline = await func(*args, **kwargs)
+    pin = Pin.get_from(instance)
+    if pin:
+        pin.onto(pipeline)
+    return pipeline
+
+
+async def traced_execute_pipeline(func, instance, args, kwargs):
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return await func(*args, **kwargs)
+
+    cmds = [format_command_args(c) for c, _ in instance.command_stack]
+    resource = "\n".join(cmds)
+    with _trace_redis_execute_pipeline(pin, config.aioredis, resource, instance):
+        return await func(*args, **kwargs)
+
+
+def traced_13_pipeline(func, instance, args, kwargs):
+    pipeline = func(*args, **kwargs)
+    pin = Pin.get_from(instance)
+    if pin:
+        pin.onto(pipeline)
+    return pipeline
+
+
+def traced_13_execute_command(func, instance, args, kwargs):
+    # If we have a _RedisBuffer then we are in a pipeline
+    if isinstance(instance.connection, _RedisBuffer):
+        return func(*args, **kwargs)
+
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    decoded_args = [arg.decode() if isinstance(arg, bytes) else arg for arg in args]
+    # This span is unactive in the context because it should not have a child span
+    # Each consecutive execute command should create a sister span
```

Comment:
Hmm what you have is correct but is slightly missing the point. Something like

> Don't activate the span since this operation is performed as a future which concludes sometime later on in execution so subsequent operations in the stack are not necessarily semantically related (we don't want this span to be the parent of all other spans created before the future is resolved).

Would be more accurate.